compiler/rustc_middle/src/ty/util.rs

```diff
@@ -850,18 +850,7 @@ impl<'tcx> OpaqueTypeExpander<'tcx> {
         }
         let args = args.fold_with(self);
         if !self.check_recursion || self.seen_opaque_tys.insert(def_id) {
-            let expanded_ty = match self.expanded_cache.get(&(def_id, args)) {
-                Some(expanded_ty) => *expanded_ty,
-                None => {
-                    for bty in self.tcx.coroutine_hidden_types(def_id) {
-                        let hidden_ty = bty.instantiate(self.tcx, args);
-                        self.fold_ty(hidden_ty);
-                    }
-                    let expanded_ty = Ty::new_coroutine_witness(self.tcx, def_id, args);
-                    self.expanded_cache.insert((def_id, args), expanded_ty);
-                    expanded_ty
-                }
-            };
+            let expanded_ty = Ty::new_coroutine_witness(self.tcx, def_id, args);
```

Comment:
I would expect us to completely remove this function 🤔 `expanded_ty` is the same type as the input and I don't think there's any use to return `None` if there's any recursion 🤔

given that we don't expand coroutine witnesses anymore, we already handly cycles involving coroutines via the ordinary occurs check. Any cyclic coroutine would have to result in a type of infinite size.

Reply:
Yes, but it's cheaper to only do this check once on the "normal" path, and only compute the layout of the coroutine if it's detected to be cyclic.

See the perf regression up-thread and the PR that fixed it.